File: elixir/protein-translation/lib/protein_translation.ex
defmodule ProteinTranslation do
  @doc """
  Given an RNA string, return a list of proteins specified by codons, in order.
  """
  @spec of_rna(String.t()) :: {atom, list(String.t())}
  def of_rna(rna) do
    rna
    |> String.to_charlist
    |> Enum.chunk_every(3)
    |> Enum.map(&codon_map/1)


  end

  @doc """
  Given a codon, return the corresponding protein

  UGU -> Cysteine
  UGC -> Cysteine
  UUA -> Leucine
  UUG -> Leucine
  AUG -> Methionine
  UUU -> Phenylalanine
  UUC -> Phenylalanine
  UCU -> Serine
  UCC -> Serine
  UCA -> Serine
  UCG -> Serine
  UGG -> Tryptophan
  UAU -> Tyrosine
  UAC -> Tyrosine
  UAA -> STOP
  UAG -> STOP
  UGA -> STOP
  """
  @spec of_codon(String.t()) :: {atom, String.t()}
  defp of_codon(codon) do
    return =  codon |> String.to_charlist |> codon_map
    {:ok, return}
  end

  def codon_map(codon) do
    case codon do
      'UGU' -> "Cysteine"
      'UGC' -> "Cysteine"
      'UUA' -> "Leucine"
      'UUG' -> "Leucine"
      'AUG' -> "Methionine"
      'UUU' -> "Phenylalanine"
      'UUC' -> "Phenylalanine"
      'UCU' -> "Serine"
      'UCC' -> "Serine"
      'UCA' -> "Serine"
      'UCG' -> "Serine"
      'UGG' -> "Tryptophan"
      'UAU' -> "Tyrosine"
      'UAC' -> "Tyrosine"
      'UAA' -> "STOP"
      'UAG' -> "STOP"
      'UGA' -> "STOP"
    end
  end

end

File: javascript/protein-translation/protein-translation.js
export const translate = (rna) => {
  if(!rna) return [];
  const protein = [];
  return loop(rna, protein);
};

const loop = (rna, protein) => {
  const codon = rna.substring(0,3);
  console.log(codon);
  rna = rna.substring(4);
  protein.push(lookup(codon));
  console.log(protein);
  if(rna.length < 3) { return protein; } else { loop(rna, protein); }
};

const lookup = (codon) => {
  return lookupTable[codon];
};

const lookupTable = {
  'AUG': "Methionine",
  'UUU': "Phenylalanine",
  'UUC': "Phenylalanine",
  'UUA': "Leucine", 
  'UUG': "Leucine",
  'UCU': "Serine",
  'UCC': "Serine", 
  'UCA': "Serine",
  'UCG': "Serine",
  'UAU': "Tyrosine",
  'UAC': "Tyrosine",
  'UGU': "Cysteine", 
  'UGC': "Cysteine",
  'UGG': "Tryptophan",
  'UAA': "STOP", 
  'UAG': "STOP", 
  'UGA': "STOP"
};